llvm/lib/CodeGen/WinEHPrepare.cpp

```diff
@@ -1236,6 +1257,140 @@ void WinEHPrepareImpl::verifyPreparedFunclets(Function &F) {
+          StringRef Name = Callee->getName();
+          if (Name == "llvm.seh.try.begin" || 
+              Name == "llvm.seh.try.end" ||
+              Name.contains("cleanup") ||
+              Name.contains("finally")) {
+            return true;
+          }
+        }
+      }
+      if (auto *LPad = dyn_cast<LandingPadInst>(&I)) {
+        if (LPad->isCleanup()) {
+          return true;
+        }
+      }
+    }
+  }
+  
+  return false;
+}
+
+/// Convert problematic call instructions to invoke instructions for SEH
+bool WinEHPrepareImpl::convertCallsToInvokesForSEH(Function &F) {
+  bool Changed = false;
+  SmallVector<CallInst *, 16> CallsToConvert;
+  
+  // First pass: identify calls that need to be converted
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      if (auto *CI = dyn_cast<CallInst>(&I)) {
+        if (shouldConvertCallToInvokeForSEH(CI)) {
+          CallsToConvert.push_back(CI);
+        }
+      }
+    }
+  }
+  
+  // Second pass: convert identified calls to invokes
+  for (CallInst *CI : CallsToConvert) {
+    BasicBlock *OrigBB = CI->getParent();
+    
+    // Create unwind destination with minimal landing pad
+    BasicBlock *UnwindBB = BasicBlock::Create(F.getContext(), 
+                                            "seh.call.unwind", &F);
+    
+    IRBuilder<> UnwindBuilder(UnwindBB);
+    StructType *LPadType = StructType::get(UnwindBuilder.getPtrTy(), 
+                                         UnwindBuilder.getInt32Ty());
+    LandingPadInst *LPad = UnwindBuilder.CreateLandingPad(LPadType, 0);
+    LPad->setCleanup(true);
+    UnwindBuilder.CreateResume(LPad);
+    
+    // Create continue block
+    BasicBlock *ContBB = BasicBlock::Create(F.getContext(), 
+                                          "seh.call.cont", &F);
+    
+    // Split the original block after the call instruction
+    BasicBlock::iterator SplitPoint = std::next(CI->getIterator());
+    if (SplitPoint != OrigBB->end()) {
+      // Move instructions after the call to the continue block
+      while (SplitPoint != OrigBB->end()) {
+        Instruction &InstToMove = *SplitPoint;
+        ++SplitPoint; // Advance before moving
+        InstToMove.moveBefore(*ContBB, ContBB->end());
+      }
+    }
+    
+    // Create the invoke instruction
+    IRBuilder<> CallBuilder(CI);
+    SmallVector<Value *, 8> Args(CI->args());
+
+    // Get the FunctionCallee from the call instruction
+    FunctionCallee Callee;
+    if (Function *F = CI->getCalledFunction()) {
+      // Direct call - create FunctionCallee from the function
+      Callee = FunctionCallee(F->getFunctionType(), F);
+    } else {
+      // Indirect call - create FunctionCallee from the called operand
+      Callee = FunctionCallee(CI->getFunctionType(), CI->getCalledOperand());
+    }
+
+    InvokeInst *Invoke = CallBuilder.CreateInvoke(Callee, ContBB, UnwindBB, Args);
+    
+    // Copy attributes and metadata
```

Comment:
You're also going to want to preserve the OpBundles and DebugLoc.
For an example, see https://github.com/llvm/llvm-project/blob/1b71ea411a9d36705663b1724ececbdfec7cc98c/llvm/lib/Target/X86/X86WinEHState.cpp#L458